channel/buffs.go

```diff
@@ -0,0 +1,744 @@
+	}
+	if meta.Jump > 0 {
+		bits = append(bits, BuffJump)
+	}
+	if meta.MAD > 0 {
+		bits = append(bits, BuffMagicAttack)
+	}
+	if meta.MDD > 0 {
+		bits = append(bits, BuffMagicDefense)
+	}
+	if meta.PAD > 0 {
+		bits = append(bits, BuffWeaponAttack)
+	}
+	if meta.PDD > 0 {
+		bits = append(bits, BuffWeaponDefense)
+	}
+	if len(bits) == 0 {
+		return
+	}
+
+	remainSec := int16(0)
+	if expiresAtMs > 0 {
+		now := time.Now().UnixMilli()
+		if d := expiresAtMs - now; d > 0 {
+			if d > 32767*1000 {
+				d = 32767 * 1000
+			}
+			remainSec = int16((d + 500) / 1000)
+		}
+	}
+
+	maskBytes := buildMaskBytes64(bits)
+	sourceID := -itemID
+	values := buildItemBuffTriplesWireOrder(meta, maskBytes, remainSec, sourceID)
+
+	// Send packets
+	cb.plr.send(packetPlayerGiveBuff(maskBytes, values, 0, 0))
+
+	// Track in memory and set timer
+	m := make([]byte, len(maskBytes))
+	copy(m, maskBytes)
+	cb.itemMasks[sourceID] = m
+
+	if expiresAtMs > 0 {
+		cb.expireAt[sourceID] = expiresAtMs
+		cb.scheduleExpiryLocked(sourceID, time.Until(time.UnixMilli(expiresAtMs)))
+	}
+}
+
+func (cb *CharacterBuffs) AddBuffFromCC(skillID int32, expiresAtMs int64, level byte, sinc1, sinc2 int, delay int16) {
+	if cb == nil || cb.plr == nil {
+		return
+	}
+	if skillID == 0 || level == 0 {
+		return
+	}
+	cb.check(skillID)
+
+	// Use configured per-skill bit positions -> build 8-byte mask deterministically (LSB-first).
+	bits, ok := skillBuffBits[skillID]
+	if !ok || len(bits) == 0 {
+		return
+	}
+	maskBytes := buildMaskBytes64(bits)
+
+	// Emit value triples in exactly the same mask byte/bit order.
+	values, remainSec := cb.buildBuffTriplesWireOrder(skillID, level, maskBytes, expiresAtMs)
+	if len(values) == 0 {
+		log.Printf("BUFF ABORT: no values produced for skillID=%d", skillID)
+		return
+	}
+
+	// Extra trailing byte only for combo/charges.
+	extra := byte(0)
+
+	// Send
+	cb.plr.send(packetPlayerGiveBuff(maskBytes, values, delay, extra))
+	if cb.plr.inst != nil {
+		cb.plr.inst.send(packetPlayerGiveForeignBuff(cb.plr.id, maskBytes, values, extra))
+	}
+
+	cb.activeSkillLevels[skillID] = level
+
+	// Compute authoritative expiry time
+	if remainSec > 0 {
+		cb.expireAt[skillID] = time.Now().Add(time.Duration(remainSec) * time.Second).UnixMilli()
+		cb.scheduleExpiryLocked(skillID, time.Duration(remainSec)*time.Second)
+	} else if expiresAtMs > 0 {
+		cb.expireAt[skillID] = expiresAtMs
+		d := time.Until(time.UnixMilli(expiresAtMs))
+		cb.scheduleExpiryLocked(skillID, d)
+	}
+}
+
+func (cb *CharacterBuffs) post(fn func()) {
+	if cb == nil || cb.plr == nil {
+		return
+	}
+	if cb.plr.inst != nil && cb.plr.inst.dispatch != nil {
+		cb.plr.inst.dispatch <- fn
+		return
+	}
+	fn()
+}
+
+func (cb *CharacterBuffs) scheduleExpiryLocked(skillID int32, after time.Duration) {
+	// Cancel previous
+	if t, ok := cb.expireTimers[skillID]; ok && t != nil {
+		t.Stop()
+		delete(cb.expireTimers, skillID)
+	}
+
+	if after <= 0 {
+		cb.post(func() { cb.expireBuffNow(skillID) })
```

Comment:
It would be safer to instead pass the playerID and skillID and perform a player lookup at part of the post job that way you don't have to be concerned about stale pointers.